Initial port of code style rule pages (#20792)
dotnet/docs: docs/fundamentals/code-analysis/style-rules/language-rules.md

---
title: Code style language rules

> ms.date: 09/25/2020

Reviewer: ```suggestion
description: Learn about the different code style rules for using C# and Visual Basic language constructs.
ms.date: 09/25/2020
```

ms.topic: reference
author: gewarren
ms.author: gewarren
dev_langs:
- CSharp
- VB
helpviewer_keywords:
- language code style rules [EditorConfig]
- language rules
- EditorConfig language conventions
---
# Language rules

> Code style language rules affect how various constructs of .NET programming languages are used, for example, modifiers and parentheses. They fall into the below categories:

Reviewer: ```suggestion
Code style language rules affect how various constructs of .NET programming languages, for example, modifiers and parentheses, are used. The rules fall into the following categories:
```

- [.NET style rules](#net-style-rules): Rules that apply to both C# and Visual Basic. The EditorConfig option names for these rules start with `dotnet_style_` prefix.
- [C# style rules](#c-style-rules): Rules that are specific to C# language only. The EditorConfig option names for these rules start with `csharp_style_` prefix.

> > [!TIP]

Reviewer: I would just delete this entire tip. We don't really use this pattern in the .NET docs.

Author: Sounds good. I will also delete the `code-language-picker.png` that I ported as part of porting this tip.

>
> - To see the code examples in your preferred programming language, choose it using the language picker at the top-right corner of the browser window.
>
>   ![Code language picker control](media/code-language-picker.png)
>
> - Use the **In this article** links to jump to different sections of the page.

## Option format

Options for language rules can be specified in an EditorConfig file with the following format:

> `option_name = value:severity`

Reviewer: I think it will be confusing to show this as the syntax, but then to say "but it won't work if you're not using VS". I think that would be fine in the VS docs, but now that this is in the .NET docs, we should make the acceptable syntax more visible. Customers aren't necessarily going to read the paragraphs; they're going to jump to the code blocks and copy them. As one of the commenters in the Roslyn issue you linked to said, it now requires two lines of configuration for the compilers to respect the rule, correct? One to say if the rule is preferred, and one to set the severity? If that's the case, I think we should include both of those lines here. Then in the paragraph, we can say something like "if you're using VS, the second line isn't necessary".

Author: Sure, I can add the second line. Its really unfortunate this is so messed up that any documentation here is going to be confusing, highlighting that the design itself is bad. I am going to try to push forward on https://github.com/dotnet/roslyn/issues/44201 so we can just delete mentioning/requiring the severity part in this entry.

Author: I am still thinking about this, and I think adding a second line here for `dotnet_diagnostic.ruleId.severity = severity` is actually going to add more confusion. This page is about configuring option values, and making them write 2 entries with severity values seems confusing. I don't think there is a good way to document this properly unless `option_name = value` is made a legal syntax, at which point we can leave out all severity related stuff from this page. I am wondering if it is fine to leave this here for now, and then remove it in a follow-up PR once I make the Roslyn change to support it. Does that sound reasonable @gewarren?

Reviewer: Yes, that sounds reasonable, thanks.

- **Value**: For each language rule, you specify a value that defines if or when to prefer the style. Many rules accept a value of `true` (prefer this style) or `false` (do not prefer this style). Other rules accept values such as `when_on_single_line` or `never`.
- **Severity**: The second part of the rule specifies the [severity level](../configuration-options.md#severity-level) for the rule. Severity specification as part of the above option syntax is only respected inside development IDEs, such as Visual Studio. This setting is not understood by the C# or VB compilers, hence not respected during build. Instead, to enforce code style rules on build, you should set the severity by using the rule ID based severity configuration syntax for analyzers. The syntax takes the form `dotnet_diagnostic.<rule ID>.severity = <severity>`, for example, `dotnet_diagnostic.IDE0040.severity = silent`. For more details, see this [github issue](https://github.com/dotnet/roslyn/issues/44201).

> [!TIP]
>

> > Starting in Visual Studio 2019 version 16.3, you can configure code style rules from the [Quick Actions](/visualstudio/ide/quick-actions.md) light bulb menu after a style violation occurs. For more information, see [Automatically configure code styles in Visual Studio](/visualstudio/ide/editorconfig-language-conventions#automatically-configure-code-styles-in-visual-studio)

Reviewer: ```suggestion
> Starting in Visual Studio 2019 version 16.3, you can configure code style rules from the [Quick Actions](/visualstudio/ide/quick-actions) light bulb menu after a style violation occurs. For more information, see [Automatically configure code styles in Visual Studio](/visualstudio/ide/editorconfig-language-conventions#automatically-configure-code-styles-in-visual-studio).
```

> ## .NET style rules

Author: Content from here on is mostly identical to the original doc and is the core index/rule reference.

  - dotnet\_style\_explicit\_tuple_names
  - dotnet\_style\_prefer\_inferred\_tuple_names
  - dotnet\_style\_prefer\_inferred\_anonymous\_type\_member_names
  - dotnet\_style\_prefer\_auto\_properties
  - dotnet\_style\_prefer\_conditional\_expression\_over\_assignment
  - dotnet\_style\_prefer\_conditional\_expression\_over\_return
  - dotnet\_style\_prefer\_compound\_assignment
- ["Null" checking preferences](#null-checking-preferences)
  - dotnet\_style\_coalesce_expression
  - dotnet\_style\_null_propagation
  - dotnet\_style\_prefer\_is\_null\_check\_over\_reference\_equality\_method
- [Parameter preferences](#parameter-preferences)
  - dotnet\_code\_quality\_unused\_parameters

### <a name="this-and-me"></a>"This." and "Me." qualifiers

This style rule can be applied to fields, properties, methods, or events. A value of **true** means prefer the code symbol to be prefaced with `this.` in C# or `Me.` in Visual Basic. A value of **false** means prefer the code element _not_ to be prefaced with `this.` or `Me.`.

These rules could appear in an *.editorconfig* file as follows:

```ini
# CSharp and Visual Basic code style settings:
[*.{cs,vb}]

> dotnet_style_qualification_for_field = false:suggestion

Reviewer: If we're moving away from this syntax, we should replace it either in this PR or as the individual rule articles are added.

Author: Yes, I am hoping to take up https://github.com/dotnet/roslyn/issues/44201 next so severity suffix here is optional. Once that is done, I will update all these docs to not mention severity as part of this syntax. Users can still specify this in VS, but it will not be documented, and eventually deprecated once we have a full fledged UI for EditorConfig.
